src/main/scala/akka/persistence/jdbc/util/SlickDriver.scala

```diff
@@ -17,22 +17,53 @@
 package akka.persistence.jdbc.util
 
 import javax.naming.InitialContext
+import javax.sql.DataSource
 
-import akka.persistence.jdbc.config.SlickConfiguration
+import akka.persistence.jdbc.config.{AsyncExecutorConfig, JournalConfig, SlickConfiguration}
 import com.typesafe.config.Config
+import slick.SlickException
 import slick.basic.DatabaseConfig
 import slick.jdbc.JdbcProfile
 import slick.jdbc.JdbcBackend._
+import slick.util.AsyncExecutor
 
 object SlickDriver {
   def forDriverName(config: Config): JdbcProfile =
     DatabaseConfig.forConfig[JdbcProfile]("slick", config).profile
 }
 
 object SlickDatabase {
+
   def forConfig(config: Config, slickConfiguration: SlickConfiguration): Database = {
-    val jndiName = slickConfiguration.jndiName.map(Database.forName(_, None))
-    val jndiDbName = slickConfiguration.jndiDbName.map(new InitialContext().lookup(_).asInstanceOf[Database])
-    jndiName.orElse(jndiDbName).getOrElse(Database.forConfig("slick.db", config))
+
+    val asyncExecConfig = slickConfiguration.asyncExecutorConfig
+
+    val namingContext = new InitialContext()
+
+    // builds a Slick Database from using a JNDI DataSource
+    // together with a properly configured AsyncExecutor
+    def buildDbFromJndiDatasource(name: String) = {
+      namingContext.lookup(name) match {
+        case dataSource: DataSource => Database.forDataSource(
+          ds = dataSource,
+          maxConnections = Option(asyncExecConfig.maxConnections),
+          executor = AsyncExecutor(
+            name = "AsyncExecutor.default",
+            minThreads = asyncExecConfig.minConnections,
+            maxThreads = asyncExecConfig.numThreads,
+            queueSize = asyncExecConfig.queueSize,
+            maxConnections = asyncExecConfig.maxConnections
+          )
```

Comment:
Just realised that this is wrong. This implementation is doing exactly what I described to be a problematic solution. 
See comment here:
https://github.com/dnvriend/akka-persistence-jdbc/issues/142#issuecomment-345085440

@WellingR suggestion to divide the number of connection across all `AsyncExecutor`s might work, but I believe it's error-prone. Users will have to size`maxConnections` in a non-intuitive fashion.

Reply:
I'm trying to devise a better solution for that. 

Maybe the simplest solution here is to say. Each component has its own configuration and we do check that they are not sharing JDNI resources. Users may configure three different `DataSource` and have different `AsyncExecutor`s associated and correctly sized for them. 

That means, we just keep the current configs, but we make sure that a `DataSource` is never shared. We let the plugin crash in case of sharing.

To achieve this, we must have something that is guaranteed to run a single time and before loading any of the three components. That bootstrap component will have to check if there is no `DataSource` sharing. Plus, it should read some global config and eventually configure a Slick Database and register it on JNDI.

Reply:
I don't think it should prohibit sharing data sources. I think this is the primary reason for using JNDI in the first place (e.g., https://github.com/dnvriend/akka-persistence-jdbc/issues/24). Otherwise, you have no elasticity between the connection pools in the three components. You could have a situation where, say the snapshot plugin is sitting on idle connections while the write journal is unable to obtain one.

I acknowledge that this resource concern also exists for threads in the executor. Maybe a good solution would be to ensure that there is exactly one `AsyncExecutor` for each JNDI name (maybe by also storing it in the JNDI context)?

The docs could also be updated to discuss some of these concerns and try to steer people away from `jndiName` in favour of `jndiDbName`.

Reply:
You are right about sharing JNDI resources. It makes no sense to prohibit because that's exactly the intention.

The problem with have with the current config format is that which piece has it's own `jndiName`, `maxConnections`, `numThreads`, etc. 

So, if we try to create one single `AsyncExecutor`, which properties should we use? 

We need a global config that should be used in case of sharing DS. This config must be read at bootstrap, create a properly configured Slick Database and associate it to `jndiDbName`.  Then, each component willing to use it will have to have a `jndiDbName`, but not a `jndiName`.